### ISSUE TITLE
h5::append not working as intended

### ISSUE THREAD

**Issue by barcode** [steven-varga/h5cpp#32]
TL;DR

* It seems h5::fill_value does not work as intended.
* There should be some diagnostic in case `h5::gzip` is used and one chunk is t0o small.

Long version:

I am trying to append data and it did not work.

Since i was not sure whether I am doing something wrong, I decided to check the current master for examples and tests.

There i found this problem:

    $ ./tests/test-packettable
    $ h5dump -d "/stream of struct" test.h5
    HDF5 "test.h5" {
    DATASET "/stream of struct" {
       DATATYPE  H5T_COMPOUND {
          H5T_STD_U16LE "field1";
          H5T_IEEE_F64LE "field2";
          H5T_IEEE_F32LE "field3";
          H5T_STD_I8LE "field4";
          H5T_IEEE_F64LE "field5";
          H5T_IEEE_F64LE "field6";
          H5T_IEEE_F64LE "field7";
          H5T_IEEE_F64LE "field8";
          H5T_IEEE_F64LE "field9";
       }
       DATASPACE  SIMPLE { ( 0 ) / ( H5S_UNLIMITED ) }
       DATA {
       }
    }
    }

The test `test-packettable` creates an empty dataset `/stream of struct` even though it writes 200 elements.

I looked at the example `examples-packettable`, since it also uses append.
In this example I looked at the dataset `/stream of integral` which is created with:

    h5::pt_t pt = h5::create<int>(fd, "stream of integral",
        h5::max_dims{H5S_UNLIMITED,3,5}, h5::chunk{2,3,5} | h5::gzip{9} | h5::fill_value<int>(3) );

If I dump it, I get this:

    $ ./examples/examples-packettable
    $ h5dump -d "/stream of integral" example.h5 
    HDF5 "example.h5" {
    DATASET "/stream of integral" {
       DATATYPE  H5T_STD_I32LE
       DATASPACE  SIMPLE { ( 6, 3, 5 ) / ( H5S_UNLIMITED, 3, 5 ) }
       DATA {
       (0,0,0): 1, 2, 3, 4, 5,
       (0,1,0): 6, 7, 8, 9, 10,
       (0,2,0): 11, 12, 13, 14, 15,
       (1,0,0): 16, 17, 18, 19, 20,
       (1,1,0): 21, 22, 23, 24, 25,
       (1,2,0): 26, 27, 28, 29, 30,
       (2,0,0): 31, 32, 33, 34, 35,
       (2,1,0): 36, 37, 38, 39, 40,
       (2,2,0): 41, 42, 43, 44, 45,
       (3,0,0): 46, 47, 48, 49, 50,
       (3,1,0): 51, 52, 53, 54, 55,
       (3,2,0): 56, 57, 58, 59, 60,
       (4,0,0): 61, 62, 63, 64, 65,
       (4,1,0): 66, 67, 68, 69, 70,
       (4,2,0): 71, 72, 73, 74, 75,
       (5,0,0): 76, 77, 78, 79, 80,
       (5,1,0): 81, 82, 83, 0, 0,
       (5,2,0): 0, 0, 0, 0, 0
       }
    }
    }

Here the values are written, but the fill value is 0 instead of 3,

Since my data is one dimensional, I added this code to the example: 

    try { // centrally used error handling
        std::vector<int> stream(83);
        std::iota(std::begin(stream), std::end(stream), 1);
        // the leading dimension is extended once chunk is full, chunk is filled in row major order
        // zero copy writes directly to chunk buffer then pushed through filter chain if specified
        // works up to H5CPP_MAX_RANK default to 7
        // last chunk if partial filled with h5::fill_value<T>( some_value )  
		h5::pt_t pt = h5::create<int>(fd, "stream of integral2",
                                      h5::max_dims{H5S_UNLIMITED}, h5::chunk{2} | h5::gzip{9} | h5::fill_value<int>(3) );
        for( auto record : stream )
            h5::append(pt, record);
    } catch ( const h5::error::any& e ){
        std::cerr << "ERROR:" << e.what();
    }

It is a copy of `/stream of integral`, but I changed the dimensions and chunk size (and name).

    $ ./examples/examples-packettable
    $ h5dump -d "/stream of integral2" example.h5 
    HDF5 "example.h5" {
    DATASET "/stream of integral2" {
       DATATYPE  H5T_STD_I32LE
       DATASPACE  SIMPLE { ( 83 ) / ( H5S_UNLIMITED ) }
       DATA {h5dump error: unable to print data

       }
    }
    }

Dumping it, I get corrupted data (but the dimensions are correct).

I got the same results with: 

* `h5::max_dims{H5S_UNLIMITED    }, h5::chunk{1    }` 
* `h5::max_dims{H5S_UNLIMITED,1  }, h5::chunk{2,1  }` 
* `h5::max_dims{H5S_UNLIMITED,2  }, h5::chunk{2,2  }` 
* `h5::max_dims{H5S_UNLIMITED,1,1}, h5::chunk{2,1,1}`
* `h5::max_dims{H5S_UNLIMITED,2,1}, h5::chunk{2,1,1}`
* `h5::max_dims{H5S_UNLIMITED,2,1}, h5::chunk{2,2,1}`
* `h5::max_dims{H5S_UNLIMITED,2,2}, h5::chunk{2,2,1}`
* `h5::max_dims{H5S_UNLIMITED,1,2}, h5::chunk{2,1,2}`
* `h5::max_dims{H5S_UNLIMITED    }, h5::chunk{5    }`

It worked with (correct values, but wrong fill value):
* `h5::max_dims{H5S_UNLIMITED,2,2  }, h5::chunk{2,2,2  }`
* `h5::max_dims{H5S_UNLIMITED,2,2,2}, h5::chunk{2,2,2,2}`
* `h5::max_dims{H5S_UNLIMITED,2,2,1}, h5::chunk{2,2,2,1}`
* `h5::max_dims{H5S_UNLIMITED,8    }, h5::chunk{2,8    }`
* `h5::max_dims{H5S_UNLIMITED      }, h5::chunk{16     }`
* `h5::max_dims{H5S_UNLIMITED      }, h5::chunk{ 8     }`
* `h5::max_dims{H5S_UNLIMITED      }, h5::chunk{ 6     }`

If I remove `h5::gzip{9}`, i can use `h5::max_dims{H5S_UNLIMITED}, h5::chunk{2}`.
Hence my guess is: This problem is caused by the minimum gzip block size (or at least the minimum size in the implementation used by hdf5).

I think some diagnostic would be good, since this error is not easy to track down.

**Comment by steven-varga** [comment]
Hi **the bug is confirmed**.  Thanks for spotting and reporting it, will take a look at it as soon as I can. This error could be the result of reworking the code to allow custom filter chain, and extending `h5::append` from rank zero to higher ranks and to other object types.

There are some minimal diagnostics already added to `H5cout.hpp` file. -- and some are buried in alternative code  However `H5diag.hpp` may be more descriptive? -- in any event if you wish to contribute then looking at the files you get the general idea, or shoot me an email if you have a better one that is simple to use and requires no dependencies?

**Comment by steven-varga** [comment]
**Wrongly applied chunk size and gzip filter combination.**

It is pedantic to apply a filter below a certain block size. On my installation the gzip filter fails below size 6. In **your example I see 4 bytes**. 

As for the default value: I added the code path, then again I don't see how it can be applied in real life. The high rank versions of append has been added to able to record movie streams frame by frame. In this setup `fill_value` has no interpretation.  A **frame must be whole | full**. 
In case of rank 1 stream, there should not be `fill_values` either; it is a stream, a sequence of data. The `END` contains information: no more data.
